### ISSUE TITLE
copy() listening too soon

### ISSUE THREAD

**Issue by beefeater7** [hluk/CopyQ#1438]
**Describe the bug**

´copy()´ is [still malfunctioning](https://github.com/hluk/CopyQ/issues/1437) in the main window. I don't know the cause (forget [this](https://github.com/hluk/CopyQ/issues/1435#issuecomment-657199935)), but it caused me to experiment and found another issue with the function. Using this script:
```
[Command]
Name=Copy Repeater
Command="
    copyq:
    disable()
    
    function popupper() {
        try { 
            copy()
            popup(clipboard())
        } catch(e) { popup(e) }
    }
    
    for (i=0; i<10; i++) {
        popupper()
    }
    
    enable()"
IsGlobalShortcut=true
Icon=\xf15b
GlobalShortcut=shift+f4
```
1. Open Notepad
2. Type "`abc`"
3. Select it; hold shift+f4 ~5sec
4. Error
5. Keep holding: more errors
6. Release: copy werks

Thanks to AHK I saw that `copy()` does wait for shortcut release to send Ctrl-C. The clipboard listener does not wait however, thus, error.

I know the case is way unlikely, but it may pertain to the main window `copy()` problem, who knows?

**Expected behavior**
Wait for keys release to listen to clipboard: cause and effect.

**Screenshots**
![image](https://user-images.githubusercontent.com/10468970/87259170-9e258e80-c4a9-11ea-9449-b919b40963ca.png)

**Version, OS and Environment**
```
CopyQ Clipboard Manager v3.12.0
Qt: 5.13.2
Compiler: GCC
Arch: x86_64-little_endian-llp64
OS: Windows 10 (10.0)
```

**Comment by pbodnar** [comment]
Oh, it looks like v3.12.0 is really broken. I made the same test as @beefeater7 and I even saw tens of CopyQ processes being created and then executed - so the activated command was not launched just once, but many times. This is definitely a new bug - in previous versions, the command was launched just once.

Apart from that, I can confirm that both versions of CopyQ seem to listen rather for _keydown_ than for _keypressed_ or _keyup_ as one would probably expect. I have no idea though if this is some enforced limitation given "by design" and whether it can harm the functionality in some way.

**Comment by hluk** [comment]
> hold shift+f4 ~5sec

This will run the command multiple times (I guess this depends on auto-repeat in system). I'm not sure what are you trying to achieve by holding the shortcut for long time, but `copy()` will fail in most cases because it either times out (since the shift key is still down) or it fails because multiple processes are trying to set clipboard simultaneously.

**Comment by beefeater7** [comment]
> Oh, it looks like v3.12.0 is really broken. I made the same test as @beefeater7 and I even saw tens of CopyQ processes being created and then executed - so **the activated command was not launched just once, but many times**. This is definitely a new bug - in previous versions, the command was launched just once.

> This will run the command multiple times (I guess this depends on auto-repeat in system).

Interesting! I disabled auto-repeat long ago, but having it on with this script kills CopyQ when holding down the keys. I just assumed it would run once for everyone.

>I'm not sure what are you trying to achieve by holding the shortcut for long time

I'm trying to break the program, I guess. The point is, the current `copy()` sequence:
`Keydown -> Listener -> Keyup -> Ctrl+C`
...opens up to this error. Rather:
`Keydown -> Keyup -> Listener -> Ctrl+C`

...but this falls outside intended use, and the auto-repeat thing makes this bug even more unusual. Auto-repeating hotkeys seem like a bigger problem, but again, only when trying to break the program.

**Comment by hluk** [comment]
This doesn't seem to me like a bug in CopyQ. I would need to implement some workaround (on all supported platforms?) to ignore listening to auto-repeated shortcut events.

I never had this problem with global shortcuts in CopyQ and don't really see a reason to change the behavior.

**Comment by pbodnar** [comment]
> This doesn't seem to me like a bug in CopyQ. I would need to implement some workaround (on all supported platforms?) to ignore listening to auto-repeated shortcut events.
>
> I never had this problem with global shortcuts in CopyQ and don't really see a reason to change the behavior.

The thing is that I didn't have this problem as well, until with the latest CopyQ version. So unless a meaningful resolution is given (and described), I wouldn't close this issue that quickly. Until then I need to stick to the previous version, because the new one is simply not usable with the new shortcuts listener behavior.

I'm also not quite sure whether forcing users to disable "the auto-repeat thing" is really the best way to go for CopyQ...

**Comment by pbodnar** [comment]
PS: I will maybe retest it once again, but I really got the feeling like the command got repeatedly called even after pressing the shortcut for a little while. Anyway, the change in behavior was quite apparent for me...

**Comment by hluk** [comment]
The behavior for global shortcuts wasn't changed for a while.

I think previously it was not an issue because either the `copy()` sent a key release events stopping the shortcut from triggering again or the `copy()` operation happened fast enough that it has already finished when the process was triggered again (which still holding the shortcut). But now the shortcut auto-repeat triggers multiple processes all waiting on `copy()` until modifier is released and then setting and checking the clipboard all at once.

**Comment by pbodnar** [comment]
@hluk, thanks for the insight. I've just tested this auto-repeat triggering again and calling `copy()` really seems to stop the shortcut from triggering again in (or until) v3.11.1. And if I comment out the call to `copy()`, the given command really gets launched multiple times in a bit uncontrollable fashion _in either version_.

So to conclude, I think this is not a real "blocker" in the end - one usually releases a keyboard shortcut quickly enough so that auto-repeat is not triggered. OTOH it would be interesting to know if the shortcuts handling couldn't be changed, so that it would wait for keyup / keypressed (and not just for keydown), provided it would make sense and it would help to eliminate the various problems observed currently or in the future...
